### PR TITLE
global: researcher profiles

### PR DESCRIPTION
**Pull request by jainaman224**

### CODE REVIEW

**Comment by lnielsen** [comment]
@slint @jainaman224 Can we take a review of the status in the coming days? @jainaman224 Would you mind preparing some screenshots with the current status of the pages.

Also, we need to have inveniosoftware/invenio-userprofiles#85 finished up, otherwise the deployment is going to be held up by not having it merged. Currently the PR is missing:

- [ ] Tests
- [ ] Simple usage documentation or as minimum integrate an example of profile customisation in the example application (https://invenio-userprofiles.readthedocs.io/en/latest/examplesapp.html)

Invenio-UserProfiles is already in beta now prior to Invenio v3 release, so unfortunately the bar for having the PR merged is higher than usual.

**Comment by jainaman224** [comment]
@lnielsen Updated screenshots of pages are at https://github.com/zenodo/zenodo/issues/1212. I will add more of them with github and orcid.
I think for `invenio-userprofiles` tests are already. Still, if you have any other testing in mind. Please tell me.
I will be added documentation by weekend.

File: zenodo/modules/profiles/ext.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            self.init_app(app)
+
+    def init_app(self, app):
+        """Flask application initialization."""
+        self.init_config(app)
+
+    def init_config(self, app):
+        """Flask application initialization."""
+        for k in dir(config):
+            if k.startswith("PROFILES_"):
```
Comment (slint) [inline zenodo/modules/profiles/ext.py]: It's better to use `ZENODO_PROFILES_` as a prefix in order to avoid collision with `invenio-userprofiles` and be more explicit about the extension

File: zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+{% assets "zenodo_search_js" %}<script src="{{ ASSET_URL }}"></script>{% endassets %}
+<script>
+$('#orcid_profile, #researcher_name, #researcher_orcid, #researcher_bio, #researcher_address, #researcher_affiliation').hide();
+
+$.ajax({
+  url: 'https://pub.orcid.org/v2.1/{{orcid_id}}/activities',
+  headers: {'Accept': 'application/json'},
+  contentType: 'application/json',
+  dataType: 'json',
+  success: function(result){
```
Comment (slint) [inline zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html]: It's better to extract this logic into some separate function, since now it looks pretty complex. There are also some utility functions available I believe that might simplify many parts of the code (eg. [Lodash's `_.map`](https://stackoverflow.com/questions/35136306/what-happened-to-lodash-pluck) for extracting deeply nested values from response objects).

File: zenodo/modules/profiles/templates/zenodo_profiles/profile.html [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            <small><a href="/communities/{{ org.id }}">{{ org.title }}</a></small>
+          </h3></li>
+          {%- endfor %}
+        </ul>
+      </div>
+      {%- endif %}
+    </div>
+
+    <div class="col-sm-9 col-xs-12">
+      <h1>Popular Uploads</h1>
```
Comment (slint) [inline zenodo/modules/profiles/templates/zenodo_profiles/profile.html]: We decided to rename it to just "Uploads" since this is a bit inaccurate (these are not the "most viewed" or "most popular" records in any way)

File: zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html [hunk 4]
```diff
@@ -0,0 +1,221 @@
+{#-
```
Comment (slint) [inline zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html]: We should check if it's possible to unify these "profile" templates to some hierarchy in order to avoid this much duplicity in code. Eg. both views have a "Popular Uploads" title and some search results, this should be either extracted in some separate template or be a Jinja macro maybe...

File: zenodo/modules/profiles/templates/zenodo_profiles/profile.html [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        <div style="margin-top:18px">
+          <a class="no_hover" href="/profile/{{ user.id }}/contact">
+            <button type="button" class="btn btn-default btn-lg btn-block">Contact</button>
+          </a>
+        </div>
+        {%- endif %}
+      {%- else %}
+      <div style="margin-top:18px">
+        <a class="no_hover" href="/account/settings/profile">
+          <button type="button" class="btn btn-default">Edit Profile</button>
```
Comment (slint) [inline zenodo/modules/profiles/templates/zenodo_profiles/profile.html]: This button can probably be removed, since the user can already click on the top right of the page on his email address to go to his Profile's settings.

File: zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+-#}
+
+{% extends "zenodo_theme/page.html" %}
+
+{%- block page_body %}
+<div class="container">
+  <div class="row" id="orcid-profile">
+    <div class="col-sm-3 col-xs-12" id="profile" style="margin-bottom:20px">
+      <br />
+      <img height="400px" width="400px"
```
Comment (slint) [inline zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html]: Some padding around the image would be useful, since now it takes way too much space compared to the rest of the page and is not as important in our use-case compared to eg. a Facebook or Twitter profile.

File: zenodo/modules/profiles/templates/zenodo_profiles/search.html [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+{% set search_hidden_params = {"q": "owners:" + user.id|string } %}
+
+{%- block page_header %}
+{% set search_input_placeholder = "Search " + user.profile.username %}
+{% include "invenio_search_ui/header.html" %}
+{%- endblock page_header %}
+
+{% block page_body %}
+{%- set global_vars = {'orcid_id': ''} %}
```
Comment (slint) [inline zenodo/modules/profiles/templates/zenodo_profiles/search.html]: It seems strange that this is not available in this scope, since it's passed from the template. Can you take a look and check if it's possible to remove usage of a  `global_vars` helper variable?

File: zenodo/modules/profiles/views.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+blueprint = Blueprint(
+    'zenodo_profiles',
+    __name__,
+    template_folder='templates',
+    static_folder='static',
+)
+
+
+@blueprint.route(
```
Comment (slint) [inline zenodo/modules/profiles/views.py]: You can move this view handler below the base `/profile/<int:owner_id>` definition, since it's basically a "sub-resource".

File: zenodo/modules/profiles/views.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        for account in user.remote_accounts:
+            if account.client_id == current_app.config.get(
+               'ORCID_APP_CREDENTIALS').get('consumer_key'):
+                orcid_id = account.extra_data.get('orcid')
+
+    email = user.email
+    default = "https://ideaclicks.in/userimages/default_user.jpg"
+    size = 400
+
+    gravatar_url = 'https://www.gravatar.com/avatar/' + hashlib.md5(
```
Comment (slint) [inline zenodo/modules/profiles/views.py]: It might be more readable to extract this as a uitlity function in `utils.py`.

File: zenodo/modules/profiles/views.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+                researcher_profile.bio = form.bio.data
+                researcher_profile.affiliation = form.affiliation.data
+                researcher_profile.location = form.location.data
+                researcher_profile.website = form.website.data
+                researcher_profile.allow_contact_owner = \
+                    form.allow_contact_owner.data
+            db.session.add(researcher_profile)
+
+            # Update email
+            if current_app.config['USERPROFILES_EMAIL_ENABLED'] and \
```
Comment (slint) [inline zenodo/modules/profiles/views.py]: I think this clause is not required since Zenodo profiles have email enabled in config

File: zenodo/modules/theme/static/scss/profiles.scss [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  @keyframes spin {
+      0% { transform: rotate(0deg); }
+      100% { transform: rotate(360deg); }
+  }
+}
+
+#researcher-profile, #orcid-profile{
+  @media screen and (max-width: 768px) {
```
Comment (slint) [inline zenodo/modules/theme/static/scss/profiles.scss]: If possible, extending some existing Bootstrap SASS class might be more clear, since manually handling CSS media queries might become tricky (and is already handled pretty well on Bootstrap's side)

File: zenodo/modules/utils/files.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
```
Comment (slint) [inline zenodo/modules/utils/files.py]: Some thing went wrong with rebasing here (this file shouldn't be part of this PR)

Comment (jainaman224) [inline zenodo/modules/utils/files.py]: Actually, It is the part of your commit as `zenodo/researcher_profile` branch is behind master.

File: zenodo/modules/profiles/views.py (hunk 10 above)
Comment (jainaman224) [inline zenodo/modules/profiles/views.py]: I think we can optimise here as well?

File: zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html (hunk 4 above)
Comment (jainaman224) [inline zenodo/modules/profiles/templates/zenodo_profiles/orcid_profile.html]: I have this template https://github.com/jainaman224/zenodo/blob/researcher_profile/zenodo/modules/profiles/templates/zenodo_profiles/results.html. If required I can include heading etc or maybe make it macro.

File: zenodo/modules/profiles/templates/zenodo_profiles/search.html (hunk 7 above)
Comment (jainaman224) [inline zenodo/modules/profiles/templates/zenodo_profiles/search.html]: How about passing it as a parameter from views?

**Comment by slint** [comment]
Rebased and merged manually

**Comment by jainaman224** [comment]
Thank you

**Comment by ccamara** [comment]
Hello, I see that this PR mas merged long time ago, but haven't been able to figure out how to create a researcher profile. Can anyone update on the status of this long-awaited feature?
